src/postprocess/refine.jl

```diff
@@ -0,0 +1,231 @@
+#
+# [CDKL2022]:
+#     E. Cancès, G. Dusson, G. Kemlin, and A. Levitt
+#     *Practical error bounds for properties in plane-wave electronic structure calculations*
+#     [SIAM Journal on Scientific Computing 44 (5), B1312-B1340](https://doi.org/10.1137/21M1456224)
+
+"""
+Invert the metric operator M.
+"""
+function invert_refinement_metric(basis::PlaneWaveBasis{T}, ψ, res) where {T}
+    # Apply the M_n operator.
+    function apply_M!(ψk, Pk, δψnk, n)
+        proj_tangent_kpt!(δψnk, ψk)
+        δψnk = sqrt.(Pk.mean_kin[n] .+ Pk.kin) .* δψnk
+        proj_tangent_kpt!(δψnk, ψk)
+        δψnk = sqrt.(Pk.mean_kin[n] .+ Pk.kin) .* δψnk
+        proj_tangent_kpt!(δψnk, ψk)
+    end
+
+    # Apply the M_n^{-1} operator.
+    function apply_inv_M(ψk, Pk, resk, n)
+        resk = proj_tangent_kpt(resk, ψk)
+        op(x) = apply_M!(ψk, Pk, x, n)
+        function f_ldiv!(x, y)
+            x .= proj_tangent_kpt(y, ψk)
+            x ./= (Pk.mean_kin[n] .+ Pk.kin)
+            proj_tangent_kpt!(x, ψk)
+        end
+        J = LinearMap{eltype(ψk)}(op, size(resk, 1))
+        # This CG seems to converge very quickly in practice (often 1 iteration).
+        cginfo = cg(J, resk;
+                    precon=FunctionPreconditioner(f_ldiv!),
+                    tol=100*eps(T), maxiter=20)
+        if !cginfo.converged
+            @warn """CG for `invert_refinement_metric` did not converge, this is unexpected.
+                    Residual norm after $(cginfo.n_iter) iterations is $(cginfo.residual_norm)."""
+        end
+        proj_tangent_kpt!(cginfo.x, ψk)
+    end
+
+    map(basis.kpoints, ψ, res) do kpt, ψk, resk
+        P = PreconditionerTPA(basis, kpt)
+        precondprep!(P, ψk)
+        δψk = similar(resk)
+        for n = 1:size(resk, 2)
+            # Apply M_n^{-1} to each band.
+            δψk[:, n] = @views apply_inv_M(ψk, P, resk[:, n], n)
+        end
+        δψk
+    end
+end
+
+"""
+Result of calling the [`refine_scfres`](@ref) function.
+
+- `basis`: Refinement basis, larger than the basis used to
+           run a first [`self_consistent_field`](@ref) computation.
+- `ψ`, `ρ`, `occupation`: Quantities from the scfres, transferred to the refinement basis
+                          and with virtual orbitals removed.
+- `δψ`, `δρ`: First order corrections to the wavefunctions and density.
+              The refined quantities are ψ + δψ and ρ + δρ.
+- `ΩpK_res`: Additional information returned by the inversion of (Ω+K)_11.
+"""
+struct RefinementResult
+    basis::PlaneWaveBasis
+    ψ
+    ρ
+    occupation
+    δψ
+    δρ
+    ΩpK_res
+end
+
+"""
+Default callback function for `refine_scfres`,
+which prints a convergence table.
+"""
+struct RefinementDefaultCallback
+    prev_time::Ref{UInt64}
+end
+function RefinementDefaultCallback()
+    RefinementDefaultCallback(Ref(zero(UInt64)))
+end
+function (cb::RefinementDefaultCallback)(info)
```

Comment:
There should be a check for the MPI master process.

Reply:
Still needed ?